components/private_channel/rust/ffi/private-channel/repsys-crypto/src/lib.rs

```diff
@@ -0,0 +1,306 @@
+use curve25519_dalek::ristretto::RistrettoPoint;
+use curve25519_dalek::scalar::Scalar;
+use rand_core::OsRng;
+
+use elgamal_ristretto::public::PublicKey;
+use elgamal_ristretto::private::SecretKey;
+use elgamal_ristretto::ciphertext::Ciphertext;
+use zkp::{CompactProof, Transcript};
+
+/// Generates asymmetric key pair and proof of knowledge of the secret (zero knowledge)
+pub fn generate_keys() -> (SecretKey, PublicKey) {
+    let sk = SecretKey::new(&mut OsRng);
+    let pk = PublicKey::from(&sk);
+    (sk, pk)
+}
+
+/// Combines two public keys to generate one where the two private keys are required to decrypt
+/// a ciphertext.
+pub fn combine_pks(pk1: PublicKey, pk2: PublicKey) -> PublicKey {
+    PublicKey::from(pk1.get_point() + pk2.get_point())
+}
+
+/// Computes an encrypted vector in which each element contains an encrypted
+/// element resulting from subtracting the encrypted hash value and the
+/// original check value
+pub fn compute_checks(
+    shared_pk: &PublicKey,
+    encrypted_hashes: Vec<Ciphertext>,
+    vector_checks: Vec<Scalar>,
+) -> Result<Vec<Ciphertext>, &'static str> {
+    if encrypted_hashes.len() != vector_checks.len() {
+        println!("Error: Size of encrypted hashes slice must be the same as vector checks (encHash: {} Vs vecCheck: {})", encrypted_hashes.len(), vector_checks.len());
+        return Err("Size of encrypted hashes slice must be the same as vector checks");
+    }
+    let encrypted_checks = vector_checks
+        .into_iter()
+        .zip(encrypted_hashes.into_iter())
+        .map(|(checks, hashes)| shared_pk.encrypt(&(&checks * &RISTRETTO_BASEPOINT_TABLE)) - hashes)
+        .collect();
+    Ok(encrypted_checks)
+}
+
+// Generate macro for Discrete Logarithm Equality proof (from zkp crate)
+define_proof! {dleq, "DLEQ Proof", (x), (A, B, H), (G) : A = (x * B), H = (x * G)}
+
+/// Randomizes and proofs correctness of operation
+pub fn randomize_and_prove(
+    encrypted_checks: &Vec<Ciphertext>,
+) -> (Vec<Ciphertext>, Vec<CompactProof>) {
+    let mut randomized_vector: Vec<Ciphertext> = Vec::new();
+    let mut proofs_correct_randomization: Vec<CompactProof> = Vec::new();
+
+    for i in 0..encrypted_checks.len() {
+        let randomizer = Scalar::random(&mut OsRng);
+
+        randomized_vector.push(
+            encrypted_checks[i] * randomizer
+        );
+        let mut transcript = Transcript::new(b"CorrectRandomization");
+        let (proof, _) = dleq::prove_compact(
+            &mut transcript,
+            dleq::ProveAssignments {
+                x: &randomizer,
+                A: &randomized_vector[i].points.0,
+                B: &encrypted_checks[i].points.0,
+                H: &randomized_vector[i].points.1,
+                G: &encrypted_checks[i].points.1,
+            },
+        );
+        proofs_correct_randomization.push(proof);
+    }
+    (randomized_vector, proofs_correct_randomization)
+}
+
+/// Verify Randomization proofs
+pub fn verify_randomization_proofs(
```

Comment:
The idea here is that the user randomises the encrypted plaintext, rather than the ciphertext itself (this is why, to randomise, we multiply by a scalar rather than adding a random point). 
If the initial plaintext is a zero (the check did not pass), then this randomisation will have no effect. In the case the initial plaintext is not equal to zero (the check did pass, or the server is trying to retrieve information of the user data), the randomisation hides the decrypted text. 
I hope this clarifies it.